## 4.9 Intro to Data Visualization with Python (TASK PART 1)

### The script contain the following points:

3A)Importing libraries to computer (Pandas, NumPy, OS)

3B)Importing 'customers' dataset 

4)Wrangle the data so that it follows consistent logic

5)Complete the fundamental data quality and consistency checks

6)Combine your customer data with the rest of your prepared Instacart data

8)Export this new dataframe as a pickle file

## 03A. Importing libraries to computer (Pandas, NumPy, OS)

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

## 03B. Importing 'customers' dataset 

In [2]:
# Creating a data importing shortcut called 'path' for the Instacart folder to make importing data easier and cleaner
path =r'/Users/drewsmith/Desktop/01-2023 Instacart Basket Analysis'

In [3]:
# Utilizing the newly created 'path' shortcut to import the 'customers' dataset
df_cus = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

## 04. Wrangle the data so that it follows consistent logic

In [4]:
# Looking if any column names need to be changed or if any columns need to be removed
df_cus.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
# Seeing how many rows and columns are in the dataset
df_cus.shape

(206209, 10)

In [6]:
# Renames the columns to something more understandable and readable
df_cus.rename(columns = {'First Name' : 'first_name', 'Surnam' : 'last_name', 'Gender' : 'gender', 'STATE' : 'state', 'Age' : 'age', 'n_dependants' : 'num_of_dependants', 'fam_status' : 'marriage_status'}, inplace = True)

In [7]:
# Confirms if column names changed
df_cus.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_of_dependants marriage_status  income  
0                  3         married  165665  
1                  0          single   59285  
2                  2         married   99568  
3                  0          single   42049  
4                  1         married   40374

### -Renamed column names and no columns needed to be removed

## 05. Complete the fundamental data quality and consistency checks

In [8]:
# Checking to see if descriptive statistics look reasonable for the numeric columns
df_cus.describe()

user_id            age  num_of_dependants         income
count  206209.000000  206209.000000      206209.000000  206209.000000
mean   103105.000000      49.501646           1.499823   94632.852548
std     59527.555167      18.480962           1.118433   42473.786988
min         1.000000      18.000000           0.000000   25903.000000
25%     51553.000000      33.000000           0.000000   59874.000000
50%    103105.000000      49.000000           1.000000   93547.000000
75%    154657.000000      66.000000           3.000000  124244.000000
max    206209.000000      81.000000           3.000000  593901.000000

In [9]:
# Checking the data type for each column and counts non-null values
df_cus.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   user_id            206209 non-null  int64 
 1   first_name         194950 non-null  object
 2   last_name          206209 non-null  object
 3   gender             206209 non-null  object
 4   state              206209 non-null  object
 5   age                206209 non-null  int64 
 6   date_joined        206209 non-null  object
 7   num_of_dependants  206209 non-null  int64 
 8   marriage_status    206209 non-null  object
 9   income             206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [10]:
# Comparing the row count to all non-null counts above to see if they match
df_cus.shape

(206209, 10)

In [11]:
# Used to find missing observations or entries in the customers dataframe
df_cus.isnull().sum()

user_id                  0
first_name           11259
last_name                0
gender                   0
state                    0
age                      0
date_joined              0
num_of_dependants        0
marriage_status          0
income                   0
dtype: int64

### -Missing values found: The first_name column is missing quite a few values (11259) since it doesn't allign with the row count in the shape function. However, nothing can be done for this since we can't assume or input anything in their place. Also, no other columns have missing values so we can utilize 'user_id' or common values in other columns to find a customer.

In [13]:
# Displays only rows that are duplicates in df_cus dataset under a new subset
df_cus_dups = df_cus[df_cus.duplicated()]

In [14]:
df_cus_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, num_of_dependants, marriage_status, income]
Index: []

### - No duplicates were found

In [15]:
# Checks for mixed types. Will display if each column is mixed or consistent
for col in df_cus.columns.tolist():
 weird = (df_cus[[col]].applymap(type) != df_cus[[col]].iloc[0].apply(type)).any(axis = 1)
 if len (df_cus[weird]) > 0:
        print (col, ' mixed')
 else: print(col, ' consistent')

user_id  consistent
first_name  mixed
last_name  consistent
gender  consistent
state  consistent
age  consistent
date_joined  consistent
num_of_dependants  consistent
marriage_status  consistent
income  consistent


In [16]:
# Changing first_name type to str
df_cus['first_name'] = df_cus['first_name'].astype('str')

In [17]:
# Now all columns are consistent
for col in df_cus.columns.tolist():
 weird = (df_cus[[col]].applymap(type) != df_cus[[col]].iloc[0].apply(type)).any(axis = 1)
 if len (df_cus[weird]) > 0:
        print (col, ' mixed')
 else: print(col, ' consistent')

user_id  consistent
first_name  consistent
last_name  consistent
gender  consistent
state  consistent
age  consistent
date_joined  consistent
num_of_dependants  consistent
marriage_status  consistent
income  consistent


### - The first_name column was mixed and then chaned to a string in order for the entire dataset to be consistent

## 06. Combine your customer data with the rest of your prepared Instacart data

In [18]:
# Utilizing the created 'path' shortcut to import the 'new_ords_prods_merge' dataset 
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'new_ords_prods_merge.pkl'))

In [19]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   

   reordered  ...    price_range_loc     busiest_day  busiest_days  \
0          0  ...  Mid-range product  Regularly busy  Regular days   
1          1  ...  Mid-range product  Regularly busy  Slowest days   
2          1  ...  Mid-range product  Regularly busy  Slowest days   
3          1  ...  Mid-range product      Least busy  Slowest days   
4          1  ...  Mid-range product      Least busy  Slowest days   

   busiest_period_of_day max_order  loyalty_flag spending_flag avg_user_price  \
0         Average orders        10  New customer   Low spender       6.367797   
1         Average orders        10  New customer   Low spender       6.367797   
2            Most orders        10  New customer   Low spender       6.367797   
3         Average orders        10  New customer   Low spender       6.367797   
4            Most orders        10  New customer   Low spender       6.367797   

  median_days_prior_order   order_frequency_flag  
0                    20.5  Non-frequent customer  
1                    20.5  Non-frequent customer  
2                    20.5  Non-frequent customer  
3                    20.5  Non-frequent customer  
4                    20.5  Non-frequent customer  

[5 rows x 25 columns]

In [20]:
df_cus.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_of_dependants marriage_status  income  
0                  3         married  165665  
1                  0          single   59285  
2                  2         married   99568  
3                  0          single   42049  
4                  1         married   40374

In [21]:
ords_prods_merge.shape

(32404859, 25)

In [22]:
df_cus.shape

(206209, 10)

In [29]:
# Merges the 'df_cus' and 'ords_prods_merge' datasets. It uses the 'user_id' column as its key. Outer join was used to not lose any data 
cus_merged1 = pd.merge(ords_prods_merge, df_cus, on = 'user_id', how = 'outer')

In [32]:
cus_merged1.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   

   reordered  ...   order_frequency_flag  first_name  last_name  gender  \
0          0  ...  Non-frequent customer       Linda     Nguyen  Female   
1          1  ...  Non-frequent customer       Linda     Nguyen  Female   
2          1  ...  Non-frequent customer       Linda     Nguyen  Female   
3          1  ...  Non-frequent customer       Linda     Nguyen  Female   
4          1  ...  Non-frequent customer       Linda     Nguyen  Female   

     state age date_joined num_of_dependants marriage_status  income  
0  Alabama  31   2/17/2019                 3         married   40423  
1  Alabama  31   2/17/2019                 3         married   40423  
2  Alabama  31   2/17/2019                 3         married   40423  
3  Alabama  31   2/17/2019                 3         married   40423  
4  Alabama  31   2/17/2019                 3         married   40423  

[5 rows x 34 columns]

In [36]:
cus_merged1.shape

(32404859, 34)

In [38]:
cus_merged1.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.240486e+07  3.240486e+07  3.240486e+07        3.240486e+07   
mean   1.710745e+06  1.029372e+05  1.714230e+01        2.738867e+00   
std    9.872988e+05  5.946610e+04  1.753532e+01        2.090077e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.559470e+05  5.142200e+04  5.000000e+00        1.000000e+00   
50%    1.711049e+06  1.026160e+05  1.100000e+01        3.000000e+00   
75%    2.565499e+06  1.543890e+05  2.400000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01        6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.240486e+07            3.032876e+07  3.240486e+07   
mean        1.342515e+01            1.110408e+01  2.559866e+04   
std         4.246380e+00            8.779064e+00  1.408400e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.354400e+04   
50%         1.300000e+01            8.000000e+00  2.530200e+04   
75%         1.600000e+01            1.500000e+01  3.794700e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered      aisle_id  department_id  \
count       3.240486e+07  3.240486e+07  3.240486e+07   3.240486e+07   
mean        8.352547e+00  5.895873e-01  7.119612e+01   9.919792e+00   
std         7.127071e+00  4.919087e-01  3.821139e+01   6.281485e+00   
min         1.000000e+00  0.000000e+00  1.000000e+00   1.000000e+00   
25%         3.000000e+00  0.000000e+00  3.100000e+01   4.000000e+00   
50%         6.000000e+00  1.000000e+00  8.300000e+01   9.000000e+00   
75%         1.100000e+01  1.000000e+00  1.070000e+02   1.600000e+01   
max         1.450000e+02  1.000000e+00  1.340000e+02   2.100000e+01   

             prices     max_order  avg_user_price  median_days_prior_order  \
count  3.240486e+07  3.240486e+07    3.240486e+07             3.240485e+07   
mean   1.198023e+01  3.305217e+01    1.198023e+01             1.039776e+01   
std    4.956554e+02  2.515525e+01    8.324227e+01             7.131754e+00   
min    1.000000e+00  1.000000e+00    1.000000e+00             0.000000e+00   
25%    4.200000e+00  1.300000e+01    7.387298e+00             6.000000e+00   
50%    7.400000e+00  2.600000e+01    7.824786e+00             8.000000e+00   
75%    1.130000e+01  4.700000e+01    8.254023e+00             1.300000e+01   
max    9.999900e+04  9.900000e+01    2.500542e+04             3.000000e+01   

                age  num_of_dependants        income  
count  3.240486e+07       3.240486e+07  3.240486e+07  
mean   4.946527e+01       1.501896e+00  9.943773e+04  
std    1.848558e+01       1.118865e+00  4.305727e+04  
min    1.800000e+01       0.000000e+00  2.590300e+04  
25%    3.300000e+01       1.000000e+00  6.700400e+04  
50%    4.900000e+01       2.000000e+00  9.661800e+04  
75%    6.500000e+01       3.000000e+00  1.279120e+05  
max    8.100000e+01       3.000000e+00  5.939010e+05

In [39]:
cus_merged1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 34 columns):
 #   Column                   Dtype   
---  ------                   -----   
 0   order_id                 int64   
 1   user_id                  int64   
 2   order_number             int64   
 3   orders_day_of_week       int64   
 4   order_hour_of_day        int64   
 5   days_since_prior_order   float64 
 6   new_customer             bool    
 7   product_id               int64   
 8   add_to_cart_order        int64   
 9   reordered                int64   
 10  product_name             object  
 11  aisle_id                 int64   
 12  department_id            int64   
 13  prices                   float64 
 14  _merge                   category
 15  price_range_loc          object  
 16  busiest_day              object  
 17  busiest_days             object  
 18  busiest_period_of_day    object  
 19  max_order                int64   
 20  loyalty_flag          

## 08. Export this new dataframe as a pickle file

In [37]:
# Export new merged file as pkl
cus_merged1.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'cus_ords_prods.pkl'))
